File: Gravels Extended Battles 1.21/common/src/main/java/drai/dev/gravelsextendedbattles/types/GravelmonTeraTypes.java
package drai.dev.gravelsextendedbattles.types;

import com.cobblemon.mod.common.api.types.*;
import com.cobblemon.mod.common.api.types.tera.*;
import drai.dev.gravelsextendedbattles.registries.*;
import net.minecraft.*;

import java.util.*;
import java.util.function.*;

public class GravelmonTeraTypes {
    public static final Map<TeraType, ChatFormatting> TERA_TYPES_TO_CHAT_FORMATTING = new HashMap<>();
    public static TeraType BLOOD;
    public static TeraType COSMIC;
    public static TeraType CRYSTAL;
    public static TeraType DIGITAL;
    public static TeraType ELDRITCH;
    public static TeraType LIGHT;
    public static TeraType QUESTIONMARK;
    public static TeraType NUCLEAR;
    public static TeraType PLASTIC;
    public static TeraType SHADOW;
    public static TeraType SLIME;
    public static TeraType SOUND;
    public static TeraType WIND;

    public static TeraType addTeraType(Function<ElementalType, TeraType> func, ElementalType elementalType, ChatFormatting chatFormatting) {
        var teraType = func.apply(elementalType);
        GravelsExtendedBattlesItems.touch();
        var shard = GravelsExtendedBattlesItems.TERA_SHARDS_BY_TYPE_NAME.get(elementalType.getName().toLowerCase());
        GravelsExtendedBattlesItems.TERA_SHARD_SUPPLIERS.put(shard, teraType);
        GravelmonTeraTypes.TERA_TYPES_TO_CHAT_FORMATTING.put(teraType, chatFormatting);
        return teraType;
    }
}

File: common/src/main/kotlin/drai/dev/gravelsextendedbattles/additions/types/tera/GravelmonTeraTypes.kt
package drai.dev.gravelsextendedbattles.additions.types.tera

import com.cobblemon.mod.common.api.types.tera.TeraType
import com.cobblemon.mod.common.api.types.tera.TeraTypes
import com.cobblemon.mod.common.api.types.tera.elemental.ElementalTypeTeraType
import com.cobblemon.mod.common.util.cobblemonResource
import drai.dev.gravelsextendedbattles.additions.types.GravelmonElementalTypes
import drai.dev.gravelsextendedbattles.mixin.invokers.TeraTypesInvoker
import net.minecraft.resources.ResourceLocation

object GravelmonTeraTypes {
    val teraTypes = TeraTypes as TeraTypesInvoker

    @JvmStatic
    val COSMIC = this.createTeraType(cobblemonResource("cosmic"), ElementalTypeTeraType(GravelmonElementalTypes.COSMIC))

    @JvmStatic
    val CRYSTAL =
        this.createTeraType(cobblemonResource("crystal"), ElementalTypeTeraType(GravelmonElementalTypes.CRYSTAL))

    @JvmStatic
    val DIGITAL =
        this.createTeraType(cobblemonResource("digital"), ElementalTypeTeraType(GravelmonElementalTypes.DIGITAL))

    @JvmStatic
    val LIGHT = this.createTeraType(cobblemonResource("light"), ElementalTypeTeraType(GravelmonElementalTypes.LIGHT))

    @JvmStatic
    val NUCLEAR =
        this.createTeraType(cobblemonResource("nuclear"), ElementalTypeTeraType(GravelmonElementalTypes.NUCLEAR))

    @JvmStatic
    val PLASTIC =
        this.createTeraType(cobblemonResource("plastic"), ElementalTypeTeraType(GravelmonElementalTypes.PLASTIC))

    @JvmStatic
    val MYSTERY =
        this.createTeraType(cobblemonResource("mystery"), ElementalTypeTeraType(GravelmonElementalTypes.MYSTERY))

    @JvmStatic
    val SHADOW = this.createTeraType(cobblemonResource("shadow"), ElementalTypeTeraType(GravelmonElementalTypes.SHADOW))

    @JvmStatic
    val SLIME = this.createTeraType(cobblemonResource("slime"), ElementalTypeTeraType(GravelmonElementalTypes.SLIME))

    @JvmStatic
    val SOUND = this.createTeraType(cobblemonResource("sound"), ElementalTypeTeraType(GravelmonElementalTypes.SOUND))

    @JvmStatic
    val WIND = this.createTeraType(cobblemonResource("wind"), ElementalTypeTeraType(GravelmonElementalTypes.WIND))

    @JvmStatic
    val ELDRITCH =
        this.createTeraType(cobblemonResource("eldritch"), ElementalTypeTeraType(GravelmonElementalTypes.ELDRITCH))

    @JvmStatic
    val BLOOD = this.createTeraType(cobblemonResource("blood"), ElementalTypeTeraType(GravelmonElementalTypes.BLOOD))

    fun createTeraType(id: ResourceLocation, type: TeraType): TeraType {
        return teraTypes.invokeCreate(id, type)
    }

    @JvmStatic
    fun touch() {
        //do nothing except call clinit
    }
}